llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -79,187 +88,241 @@ const Constant *getAliasee(const GlobalAlias *GA) {
 
   // Add the GO's dependencies to its hash.
-  Changed = updateDigestUseDependencies(GOHash, State.Dependencies, Visited,
-                                        GOIMap, AccumulateGODigest) ||
-            Changed;
-
-  return Changed;
+  const auto result = updateDigestUseDependencies(
+      GO, GOHash, GODepth, State.Dependencies, HasRepoDefinitionMD);
+  Visited.erase(GO);
+  LLVM_DEBUG(dbgs() << "Removing \"" << GO->getName()
+                    << "\" from the Visited map\n");
+  return result;
 }
 
-// Calculate the GOs' initial digest, dependencies, contributions and GONumber.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M) {
-  GOInfoMap GOIMap;
-  GONumber GONum;
-
-  for (auto &GV : M.globals()) {
-    if (GV.isDeclaration())
-      continue;
-    calculateGOInfo(&GV, GOIMap);
-    ++GONum.VarNum;
-  }
-  for (auto &Fn : M.functions()) {
-    if (Fn.isDeclaration())
-      continue;
-    calculateGOInfo(&Fn, GOIMap);
-    ++GONum.FuncNum;
-  }
-  return std::make_pair(std::move(GOIMap), std::move(GONum));
+DigestType ModuleHashGenerator::calculateDigest(const GlobalObject *GO,
+                                                bool HasRepoDefinitionMD) {
+  auto const Result = std::get<DigestType>(
+      updateDigestUseDependenciesAndContributions(GO, HasRepoDefinitionMD));
+  assert(Visited.empty() && "Visited is expected to be empty!");
+  return Result;
 }
 
-std::tuple<bool, unsigned, unsigned> generateRepoDefinitions(Module &M) {
-  bool Changed = false;
-  GONumber GONum;
-  GOStateMap Visited;
-  Visited.reserve(M.size() + M.getGlobalList().size());
-  GOInfoMap GOIMap;
+void ModuleHashGenerator::digestModule(Module &M) {
+  auto GOSize = M.size() + M.getGlobalList().size();
+  Visited.reserve(GOSize);
+  GOHashCache.reserve(GOSize);
+  Changed = false;
 
-  // Calculate the GOInfoMap and GONumber.
-  std::tie(GOIMap, GONum) = calculateGONumAndGOIMap(M);
-
-#ifndef NDEBUG
-  for (auto &GI : GOIMap) {
-    LLVM_DEBUG(dbgs() << "\nGO Name:" << GI.first->getName() << "\n");
-    LLVM_DEBUG(GI.second.dump());
-  }
-#endif
+  // Calculate the GOInfoMap.
+  calculateGONumAndGOIMap(M);
 
   // Update the GO's digest using the dependencies and the contributions.
   for (auto &GO : M.global_objects()) {
     if (GO.isDeclaration())
       continue;
-    Visited.clear();
-    MD5 Hash = MD5();
-    auto Helper = [](const GlobalObject *GO, MD5 &GOHash, GOInfoMap &GOIMap) {
-      return accumulateGOInitialDigestAndGetGODigestState(GO, GOHash, GOIMap);
-    };
-    updateDigestUseDependenciesAndContributions(&GO, Hash, Visited, GOIMap,
-                                                Helper);
-    MD5::MD5Result Digest;
-    Hash.final(Digest);
-    set(&GO, Digest);
-    Changed = true;
+    set(&GO, calculateDigest(&GO, false));
   }
+}
+
+DigestType
+ModuleHashGenerator::calculateDigestUseRepoDefinition(const GlobalObject *GO) {
+  auto const Result = std::get<DigestType>(
```

Comment:
Given that Visited is a member of this class, it could also be asserted that it is empty before starting to use it...

Reply:
Added, thanks.